aten/src/ATen/Utils.h

```diff
@@ -127,4 +128,25 @@ inline int64_t prod_intlist(ArrayRef<int64_t> list) {
   return std::accumulate(list.begin(), list.end(), 1ll, std::multiplies<int64_t>());
 }
 
+/**
+ * Utility function used in tensor implementations, which
+ * supplies the default generator to tensors, if an input generator
+ * is not supplied. The input Generator* is also dynamic casted to
+ * the backend generator type (CPU/CUDAGenerator etc.)
+ * 
+ * See Note [Thread-safety and Generators]
+ */
+template <typename T>
+static inline T * check_generator_with_default(Generator * expr, Generator * defaultValue) {
+  if (!expr) {
+    expr = defaultValue;
+  }
+  // See Note [Thread-safety and Generators]
+  std::lock_guard<std::mutex> lock(expr->mutex_);
+  if (T::device_type() == expr->device().type()) {
+    return static_cast<T*>(expr);
+  }
+  AT_ERROR("Expected a '", typeid(T).name(), "' but found '", typeid(expr).name(), "'");
```

Comment:
Nit: I am aware the old code did it this way, but it would be not to use `typeid` for the error message here, since you won't get useful information when we compile without RTTI. Just print out the device type!